doc/development/powershell.md

```diff
@@ -0,0 +1,165 @@
+
+- Reserve powershell pipelines (`|` operator) for simple assignment operations only (e.g. light transformations of a data structure)
+  - If side effects are needed, favor `foreach` loops.
+  - If the value being returned is any sort of collection (e.g. list), either favor a `foreach` loop or explicitly declare a variable type like `[array]$arr = some-func | ? { $_ -contains "foo" }` . Due to how pipelines work, when the result value is singular, the inner value with the collection item type will be returned instead of a collection with length 1. _This goes the same for powershell functions._
+- Handle exit codes from calls to external binaries (e.g. `git`).
+  - Powershell by default will only stop execution on thrown exceptions. Calls to external binaries do not throw powershell native exceptions, so failures will not be caught unless explicitly handled.
+    - Check the `$LASTEXITCODE` variable for values greater than 0.
+    - Use the new `$PSNativeCommandUseErrorActionPreference` variable coming in 7.3 (https://docs.microsoft.com/en-us/powershell/scripting/whats-new/what-s-new-in-powershell-73?view=powershell-7.2#experimental-features).
+- Set global powershell variables to achieve consistent behavior:
+  - `$ErrorActionPreference` should be set to `Stop` at the top of any cmdlet files unless there is a specific reason to do something else.
+  - `Set-StrictMode -Version 4` should be set at the top of any cmdlet files unless there is a specific reason to do something else.
+- Support the common `-WhatIf` parameter. See [SupportsShouldProcess docs](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-shouldprocess?view=powershell-7.2#supportsshouldprocess).
+  - Scripts should be able to be run locally, and they should also support a dry run mode via `-WhatIf` so a user can see what actions might be performed that could be destructive.
+  - Most common powershell module functions will support `-WhatIf` natively, but engsys code that does any write operations should also gate this functionality behind this parameter by checking [ShouldProcess](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-shouldprocess?view=powershell-7.2#pscmdletshouldprocess).
+- Log write actions with enough context that someone reading the log statement in a pipeline would be able to know what happened.
+  - For example, `Write-Host "Updating code owner to '$user' in '$ServiceDirectory'"` is a much better log than `Write-Host "Updating codeowner"`.
+
+## Style
+
+Powershell is a very flexible language. In engsys we try to use a smaller subset of what's possible to keep programs consistent and their behavior predictable.
+
+- Declare types for function parameters (e.g. `function foo([string]$bar) {}`.
```

Comment:
IMO we should highlight 

```pwsh
<#
.SYNOPSIS
<blah>

.DESCRIPTION
<blah>

.PARAMETER blah
<blah>
#>
function meep([string]$blah)
```

At either A) the entrypoint of the script or B) the "public" interface points on functions that are used elsewhere.

Reply:
Makes sense. I'm also wondering whether we should keep trying to do the markdown generation thing or not (e.g. https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/TestResources/New-TestResources.ps1.md), since I'm not sure who looks at it or if anyone else knows how to generate it.

Reply:
I'll definitely am interested. If I can figure it out I'll add it to our powershell CI.